File: binarySearch7/SearchA2DMatrix.java

package binarySearch7;

// Approach 2
//the time complexity of the binary search algorithm is O(log(row * col)),
//which is equivalent to O(log n), where n is the total number of elements in the matrix.
//Space complexity: O(1)
public class SearchA2DMatrix {
        public boolean searchMatrix(int[][] matrix, int target) {
           if(matrix.length ==0){
               return false;
           }
           int row = matrix.length;
           int col = matrix[0].length; // matrix[0] means number of elements in the 0 row is the actual coloum size.
            int left = 0;
            int right = (row*col) -1;
            while(right>=left){
                int mid = (right+left)/2;
                if(matrix[mid/col][mid%col]==target){
                    return true;
                }else if(matrix[mid/col][mid%col]>target){
                    right= mid-1;
                }else{
                    left = left+1;
                }
            }
            return false;
    }
    
    public static void main(String[] args) {
     SearchA2DMatrix obj = new SearchA2DMatrix();
        
        int [][] matrix = {{1,3,5,7},{10,11,16,20},{23,30,34,60}};
         int target = 3;
         System.out.println(obj.searchMatrix(matrix, target));
    }
}

// Approach 1 
// Brute force metthod
// Time complexity O(m*n)
// space complexity O(1)
/*
public class SearchA2DMatrix {
        public boolean searchMatrix(int[][] matrix, int target) {
           for(int i=0;i<matrix.length;i++){
               for(int j=0;j<matrix.length;j++){
                   if(matrix[i][j]==target){
                       return true;
                   }
               }
           }
           return false;
    }
    
    public static void main(String[] args) {
     SearchA2DMatrix obj = new SearchA2DMatrix();
        
        int [][] matrix = {{1,3,5,7},{10,11,16,20},{23,30,34,60}};
         int target = 3;
         System.out.println(obj.searchMatrix(matrix, target));
    }
}

*/

File: LeetCode128/Search a 2D Matrix.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Remoting.Messaging;
using System.Text;
using System.Threading.Tasks;

namespace LEETCODE128
{

    internal class Search_a_2D_Matrix
    {
        // Approach 2
        // TC O(log(m*n))
        // SC O(1)
        public bool SearchMatrix(int[][] matrix, int target)
        {
            int row = matrix.Length;
            int coloum = matrix[0].Length;
            int left = 0;
            int right = row * coloum - 1;
            while(left <= right)
            {
                int mid = left +(right - left)/2;
                if (matrix[mid/coloum][mid%coloum] == target)
                {
                    return true;
                }else if(matrix[mid / coloum][mid % coloum] > target)
                {
                    right = mid - 1;
                }
                else
                {
                    left = mid + 1;
                }
            }
            Console.WriteLine(coloum);
            return false;
        }

     /*   static void Main(string[] args)
        {
            Search_a_2D_Matrix obj = new Search_a_2D_Matrix();
            int target = 3;
            int[][] matrix = new int[][]
            {
               new int[] { 1, 3, 5, 7 },
               new int[] { 10, 11, 16, 20 },
               new int[] { 23, 30, 34, 60 }
            };
            Console.WriteLine(obj.SearchMatrix(matrix, target));

        }*/
    }


    // Approach 1 
    // TC O(m*n)
    // SC O(1)
    /*  internal class Search_a_2D_Matrix
      {

          public bool SearchMatrix(int[][] matrix, int target)
          {
              for(int i = 0;i< matrix.Length; i++)
              {
                   for(int j = 0;j< matrix[i].Length; j++)
                  {
                      if (matrix[i][j] == target)
                      {
                           return true;
                      }
                  }
              }
              return false;

          }

          static void Main(string[] args)
          {
              Search_a_2D_Matrix obj = new Search_a_2D_Matrix();
              int target = 3;
              int[][] matrix = new int[][]
              { 
                 new int[] { 1, 3, 5, 7 },
                 new int[] { 10, 11, 16, 20 }, 
                 new int[] { 23, 30, 34, 60 } 
              };
              Console.WriteLine(obj.SearchMatrix(matrix,target));

          }
      }*/

}
